flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/TimeBoundedStreamJoinOperator.java

```diff
@@ -0,0 +1,398 @@
+
+		addToLeftBuffer(leftValue, leftTs);
+
+		long min = leftTs + lowerBound;
+		long max = leftTs + upperBound;
+
+		// TODO: Adapt to different bucket sizes here
+		// Go over all buckets that are within the time bounds
+		for (long i = min; i <= max; i++) {
+			List<Tuple3<T2, Long, Boolean>> fromBucket = rightBuffer.get(i);
+			if (fromBucket != null) {
+
+				// check for each element in current bucket if it should be joined
+				for (Tuple3<T2, Long, Boolean> tuple : fromBucket) {
+					if (shouldBeJoined(leftTs, tuple.f1)) {
+
+						// collect joined tuple with left timestamp
+						collect(leftValue, tuple.f0, leftTs, tuple.f1);
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * Process a {@link StreamRecord} from the right stream. Whenever a {@link StreamRecord}
+	 * arrives at the right stream, it will get added to the right buffer. Possible join candidates
+	 * for that element will be looked up from the left buffer and if the pair lies within the user
+	 * defined boundaries, it gets collected.
+	 *
+	 * @param record An incoming record to be joined
+	 * @throws Exception Can throw an exception during state access
+	 */
+	@Override
+	public void processElement2(StreamRecord<T2> record) throws Exception {
+
+		long rightTs = record.getTimestamp();
+		T2 rightElem = record.getValue();
+
+		addToRightBuffer(rightElem, rightTs);
+
+		long min = rightTs + inverseLowerBound;
+		long max = rightTs + inverseUpperBound;
+
+		// TODO: Adapt to different bucket sizes here
+		// Go over all buckets that are within the time bounds
+		for (long i = min; i <= max; i++) {
+			List<Tuple3<T1, Long, Boolean>> fromBucket = leftBuffer.get(i);
+			if (fromBucket != null) {
+
+				// check for each element in current bucket if it should be joined
+				for (Tuple3<T1, Long, Boolean> tuple : fromBucket) {
+					if (shouldBeJoined(tuple.f1, rightTs)) {
+
+						// collect joined tuple with left timestamp
+						collect(tuple.f0, rightElem, tuple.f1, rightTs);
+					}
+				}
+			}
+		}
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+
+		// remove from both sides all those elements where the timestamp is less than the lower
+		// bound, because they are not considered for joining anymore
+		removeFromLhsUntil(mark.getTimestamp() + inverseLowerBound);
+		removeFromRhsUntil(mark.getTimestamp() + lowerBound);
+
+		if (timeServiceManager != null) {
+			timeServiceManager.advanceWatermark(mark);
+		}
+
+		output.emitWatermark(new Watermark(mark.getTimestamp() - getWatermarkDelay()));
+	}
+
+	private void collect(T1 left, T2 right, long leftTs, long rightTs) throws Exception {
+		collector.setAbsoluteTimestamp(leftTs);
+		context.leftTs = leftTs;
+		context.rightTs = rightTs;
+		userFunction.processElement(left, right, context, this.collector);
+	}
+
+	private void removeFromLhsUntil(long maxCleanup) throws Exception {
+
+		// setup state
+		if (this.lastCleanupRightBuffer.value() == null) {
+			this.lastCleanupRightBuffer.update(0L);
```

Comment:
This state update is unnecessary since it's will be updated again at the end of this method. You can just check if `lastCleanupLeft` is null in the next few lines